src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -0,0 +1,247 @@
+                MTR_LOG_INFO("%@ connectivity now viable", self);
+                [self _callHandler];
+            }
+        });
+        nw_connection_start(connection);
+    }
+}
+
+static void _resolveReplyCallback(
+    DNSServiceRef sdRef,
+    DNSServiceFlags flags,
+    uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode,
+    const char * fullname,
+    const char * hosttarget,
+    uint16_t port, /* In network byte order */
+    uint16_t txtLen,
+    const unsigned char * txtRecord,
+    void * context)
+{
+    auto * connectivityMonitor = (__bridge MTRDeviceConnectivityMonitor *) context;
+    [connectivityMonitor handleResolvedHostname:hosttarget port:port error:errorCode];
+}
+
+- (void)startMonitoringWithHandler:(MTRDeviceConnectivityMonitorHandler)handler queue:(dispatch_queue_t)queue
+{
+    std::lock_guard lock(sConnectivityMonitorLock);
+
+    MTRDeviceConnectivityMonitorHandler handlerCopy = [handler copy];
+    _monitorHandler = handlerCopy;
+    _handlerQueue = queue;
+
+    // If there's already a resolver running, just return
+    if (_resolver) {
+        MTR_LOG_INFO("%@ connectivity monitor updated handler", self);
+        return;
+    }
+
+    MTR_LOG_INFO("%@ start connectivity monitoring for %@ (%lu monitoring objects)", self, _instanceName, static_cast<unsigned long>(sConnectivityMonitorCount));
+
+    _resolver = [MTRDeviceConnectivityMonitor _sharedResolverConnection];
+    if (!_resolver) {
+        MTR_LOG_ERROR("%@ failed to get shared resolver connection", self);
+        return;
+    }
+    DNSServiceErrorType dnsError = DNSServiceResolve(&_resolver,
+        kDNSServiceFlagsShareConnection,
+        kDNSServiceInterfaceIndexAny,
+        _instanceName.UTF8String,
+        kOperationalType,
+        kLocalDot,
+        _resolveReplyCallback,
+        (__bridge void *) self);
+    if (dnsError != kDNSServiceErr_NoError) {
+        MTR_LOG_ERROR("%@ failed to create resolver", self);
+        return;
+    }
+
+    sConnectivityMonitorCount++;
+}
+
+#define MTRDEVICECONNECTIVITYMONITOR_SHARED_CONNECTION_LINGER_INTERVAL (10)
+
+- (void)_stopMonitoring
+{
+    os_unfair_lock_assert_owner(&sConnectivityMonitorLock);
+    for (NSString * hostName in _connections) {
+        nw_connection_cancel(_connections[hostName]);
+    }
+    [_connections removeAllObjects];
+
+    if (_resolver) {
+        DNSServiceRefDeallocate(_resolver);
+        _resolver = NULL;
+
+        // If no monitor objects exist, schedule to deallocate shared connection and queue
+        sConnectivityMonitorCount--;
+        if (!sConnectivityMonitorCount) {
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) (MTRDEVICECONNECTIVITYMONITOR_SHARED_CONNECTION_LINGER_INTERVAL * NSEC_PER_SEC)), sSharedResolverQueue, ^{
+                std::lock_guard lock(sConnectivityMonitorLock);
+
+                if (!sConnectivityMonitorCount) {
+                    MTR_LOG_INFO("%@ Closing shared resolver connection", self);
```

Comment:
```suggestion
                    MTR_LOG_INFO("MTRDeviceConnectivityMonitor: Closing shared resolver connection");
```
Avoid capturing `self` in the dispatch_after block.

Thought this whole thing seems half way between a premature optimization, and managing the life time of the shared connection along with other global state in the device controller factory.

Reply:
In the case of many devices getting subscribed to on reboot, and the upcoming changes to how subscriptions are scheduled, I feel that we will be running into repeated creation/dealloc of the shared connection, and so I put this in, to avoid needing to do this later in the cycle and miss testing.